sklearn/utils/validation.py

```diff
@@ -1668,3 +1668,36 @@ def _get_feature_names(X):
     # Only feature names of all strings are supported
     if types[0] == "str":
         return feature_names
+
+
+def _check_feature_names_in(estimator, input_features):
+    """Make feature_names_in_ from estimator.
+
+    Used to validate input from `get_feature_names_out(input_features=None)`.
+    """
+    feature_names_in_ = getattr(estimator, "feature_names_in_", None)
+    if input_features is not None:
+        input_features = np.asarray(input_features, dtype=object)
+        if feature_names_in_ is not None and not np.array_equal(
+            feature_names_in_, input_features
+        ):
+            raise ValueError("input_features is not equal to feature_names_in_")
+
+        if (
+            hasattr(estimator, "n_features_in_")
+            and len(input_features) != estimator.n_features_in_
+        ):
+            raise ValueError(
+                "input_features should have length equal to number of "
+                f"features ({estimator.n_features_in_}), got {len(input_features)}"
+            )
+        return input_features
+
+    if feature_names_in_ is not None:
+        return feature_names_in_
+
+    # Only generates features if `n_features_in_` is defined
+    if hasattr(estimator, "n_features_in_"):
+        return np.asarray(
+            [f"x{i}" for i in range(estimator.n_features_in_)], dtype=object
+        )
```

Comment:
Shouldn't we raise a `ValueError` with an informative error message if we reach this line? I think returning `None` is bound to cause dowstream exception with unintelligible error messages.

Reply:
Are there any transformers in scikit-learn that do not have `n_features_in_` defined for a legitimate reason where this function would be called with `input_features=None`?

Reply:
`Pipeline` with a vectorizer as the first step. Looking at this again, I think it's better for `Pipeline.get_feature_names_out` to not validate `input_features` and let the first step handle the validation. 

This means this we can error here if `n_features_in_` is not defined.